lib/backend/firestore/firestorebk.go

```diff
@@ -709,19 +709,38 @@ func (b *Backend) purgeExpiredDocuments() error {
 // deleteDocuments removes documents from firestore in batches to stay within the
 // firestore write limits
 func (b *Backend) deleteDocuments(docs []*firestore.DocumentSnapshot) error {
-	for i := 0; i < len(docs); i += commitLimit {
-		batch := b.svc.Batch()
+	// Deduplicate documents. The Firestore SDK will error if duplicates are found,
+	// but existing callers of this functionassume this is valid.
+	mDocs := make(map[string]*firestore.DocumentSnapshot)
+	for _, doc := range docs {
+		mDocs[doc.Ref.Path] = doc
+	}
 
-		for j := 0; j < commitLimit && i+j < len(docs); j++ {
-			batch.Delete(docs[i+j].Ref)
-		}
+	batch := b.svc.BulkWriter(b.clientContext)
+	jobs := make([]*firestore.BulkWriterJob, 0, len(mDocs))
 
-		if _, err := batch.Commit(b.clientContext); err != nil {
+	i := 0
+	for _, doc := range mDocs {
+		job, err := batch.Delete(doc.Ref)
+		if err != nil {
 			return ConvertGRPCError(err)
 		}
+
+		jobs = append(jobs, job)
+		i++
+		if i >= commitLimit {
+			batch.Flush()
+		}
 	}
 
-	return nil
+	var errs []error
```

Comment:
don't forget to call `batch.End` after inserting every element into the batch in order to handle the jobs added since the last flush call otherwise those won't be committed

Reply:
👍 
Thanks for the catch. Missed this since tests passed due to the commit triggering due to internal logic within the batching system (took a look, it does a lot of wierd things).